src/t_list.c

```diff
@@ -95,58 +94,70 @@ void listTypeTryConversionForValues(robj *o, robj **argv, int start, int end,
+            /* Invoke callback before conversion. */
+            if (fn) fn(data);
 
-        listTypeConvertToQuicklist(o); 
+            listTypeConvertToQuicklist(o); 
+        }
+    } else if (enc == OBJ_ENCODING_LISTPACK) {
+        /* Nothing to do... */
+    } else {
+        serverPanic("Unknown list encoding");
     }
+
 }
 
 /* Check the length and size of a quicklist to see if we need to convert it to listpack.
  *
  * If callback is given the function is called in order for caller to do some work
  * before the list conversion. */
-void listTypeTryConvertToListpack(robj *o, beforeConvertCB fn, void *data) {
-    size_t sz_limit;
-    unsigned long count_limit;
-    quicklist *ql = o->ptr;
+static void listTypeTryConvertQuicklist(robj *o, int enc, beforeConvertCB fn, void *data) {
+    serverAssert(o->encoding == OBJ_ENCODING_QUICKLIST);
+
+    if (enc == OBJ_ENCODING_QUICKLIST) {
+        /* Nothing to do... */
+    } else if (enc == OBJ_ENCODING_LISTPACK) {
+        size_t sz_limit;
+        unsigned long count_limit;
+        quicklist *ql = o->ptr;
 
-    /* A quicklist can be converted to listpack only if it
-     * has only one packed node. */
-    if (o->encoding != OBJ_ENCODING_QUICKLIST || ql->len != 1 ||
-        ql->head->container != QUICKLIST_NODE_CONTAINER_PACKED)
-    {
-        return;
-    }
+        /* A quicklist can be converted to listpack only if it
+         * has only one packed node. */
+        if (ql->len != 1 || ql->head->container != QUICKLIST_NODE_CONTAINER_PACKED)
+            return;
 
-    /* Note that to avoid frequent conversions of quicklist and listpack due to frequent
-     * insertion and deletion, we don't convert quicklist to listpack until its length or
-     * size is less than half of the limit. */
-    quicklistSizeAndCountLimit(server.list_max_listpack_size,&sz_limit,&count_limit);
-    if ((sz_limit != SIZE_MAX && ql->head->sz > sz_limit/2) ||
-        (count_limit != ULONG_MAX && ql->count > count_limit/2))
-    {
-        return;
-    }
+        /* Note that to avoid frequent conversions of quicklist and listpack due to frequent
+         * insertion and deletion, we don't convert quicklist to listpack until its length or
+         * size is less than half of the limit. */
+        quicklistSizeAndCountLimit(server.list_max_listpack_size,&sz_limit,&count_limit);
+        if ((sz_limit != SIZE_MAX && ql->head->sz > sz_limit/2) ||
+            (count_limit != ULONG_MAX && ql->count > count_limit/2))
+        {
+            return;
+        }
 
-    /* Invoke callback before conversion. */
-    if (fn) fn(data);
+        /* Invoke callback before conversion. */
+        if (fn) fn(data);
 
-    listTypeConvertToListpack(o);
+        listTypeConvertToListpack(o);
+    } else {
+        serverPanic("Unknown list encoding");
+    }
 }
 
-void listTypeTryConversion(robj *o, beforeConvertCB fn, void *data) {
+void listTypeTryConversion(robj *o, int enc, beforeConvertCB fn, void *data) {
```

Comment:
let's document that `enc` is the target encoding, and that do do that to avoid resonance.
btw, maybe instead it should be "growing" / "shrinking" boolean input?
that's how we use it, right?

Reply:
At the end i change `listTypeTryConversionForValues` to `listTypeTryConversionForGrowing` and change `listTypeTryConversion` to `listTypeTryConversionForShrinking`.
@zuiderkwast :-) you're right.